File: python/src/core/problem.py
import itertools
from dataclasses import dataclass
from typing import List, Dict, Tuple, Optional

from graphviz import Digraph


# General design notes:
# * avoid users having to subclass at any cost
# * ids are optional and only there for debugging, they don't have semantics

# TODO make ids not optional

@dataclass(eq=False, frozen=True)
class Memory:
    id: Optional[str]
    size_bits: Optional[int]


# TODO switch to single-directional channels with possibly shared bandwidth restrictions
#  then directionality can just be encoded as two separate channels
@dataclass(eq=False, frozen=True)
class Channel:
    id: Optional[str]

    memory_a: Memory
    memory_b: Memory
    dir_a_to_b: bool
    dir_b_to_a: bool

    # TODO make this calculation a dynamic function that gets both input sizes?
    #  this is going more back towards inheritance, but maybe it's worth it here
    latency: float
    time_per_bit: float
    energy_per_bit: float


@dataclass(eq=False, frozen=True)
class Core:
    id: Optional[str]

    # TODO technically this is redundant, it's only here for error checking
    #  maybe we should derive this from the allocation options?
    connected_memories: List[Memory]


def dot_table(title: str, rows: List[Tuple[str, str]]) -> str:
    table = f"<TABLE BORDER=\"0\" COLUMNS=\"*\" ROWS=\"*\"><TR><TD colspan=\"2\"><B>{title}</B></TD></TR>"
    for key, value in rows:
        table += f"<TR><TD>{key}</TD><TD>{value}</TD></TR>"
    table += "</TABLE>"
    return table


def dot_html(content: str) -> str:
    return f"<{content}>"


@dataclass(eq=False, frozen=True)
class Hardware:
    id: Optional[str]
    cores: List[Core]
    memories: List[Memory]
    channels: List[Channel]

    def assert_valid(self):
        for core in self.cores:
            for mem in core.connected_memories:
                assert mem in self.memories
        for chan in self.channels:
            assert chan.memory_a in self.memories
            assert chan.memory_b in self.memories

    def to_graphviz(self) -> Digraph:
        dot = Digraph()
        if self.id is not None:
            dot.attr(label=f"<<B>{self.id}</B>>")
            dot.attr(labelloc="t")

        for i, mem in enumerate(self.memories):
            size = "inf" if mem.size_bits is None else str(mem.size_bits)
            label = dot_html(dot_table("Memory", [("id", mem.id), ("size_bits", size)]))
            dot.node(f"mem-{i}", label, shape="box",
                     color="blue")

        for i, core in enumerate(self.cores):
            label = dot_html(dot_table("Core", [("id", core.id)]))
            dot.node(f"core-{i}", label, color="green")
            for j, mem in enumerate(core.connected_memories):
                head = f"core-{i}"
                tail = f"mem-{self.memories.index(mem)}"
                dot.edge(tail, head, headlabel=str(j), dir="none")

        for i, chan in enumerate(self.channels):
            rows = [
                ("id", chan.id),
                ("latency", str(chan.latency)),
                ("time_per_bit", str(chan.time_per_bit)),
                ("bandwidth", str(1/chan.time_per_bit)),
                ("energy_per_bit", str(chan.energy_per_bit))
            ]
            label = dot_html(dot_table("Channel", rows))
            mid = f"chan-{i}"
            dot.node(mid, label, shape="box", style="rounded", color="orange")

            head = f"mem-{self.memories.index(chan.memory_a)}"
            tail = f"mem-{self.memories.index(chan.memory_b)}"

            dirs = {(False, False): "none", (False, True): "back", (True, False): "forward", (True, True): "both"}
            dir = dirs[(chan.dir_a_to_b, chan.dir_b_to_a)]

            dot.edge(tail, mid, dir=dir)
            dot.edge(mid, head, dir=dir)

        return dot


# TODO dedicated slice, concat, transpose, pad ... operations that don't require computation,
#   but only specific memory transfers
# TODO dedicated input node
# TODO change to a mutable graph representation that can also deal with accumulation
@dataclass(eq=False, frozen=True)
class OperationNode:
    id: Optional[str]
    size_bits: int
    # no inputs means this is an input node
    inputs: List['OperationNode']


# TODO split operations and values to:
#  * support multiple outputs for a single node
#  * clearly separate input nodes from the operation nodes
@dataclass(eq=False, frozen=True)
class OperationGraph:
    id: Optional[str]
    nodes: List[OperationNode]

    inputs: List[OperationNode]
    outputs: List[OperationNode]

    def assert_valid(self):
        for node in self.nodes:
            for input in node.inputs:
                assert input in self.nodes
        assert len(set(self.inputs)) == len(self.inputs)
        assert len(set(self.outputs)) == len(self.outputs)
        for input in self.inputs:
            assert input in self.nodes
        for output in self.outputs:
            assert output in self.nodes

    def to_graphviz(self):
        dot = Digraph()

        if self.id is not None:
            dot.attr(label=f"<<B>{self.id}</B>>")
            dot.attr(labelloc="t")

        for i, node in enumerate(self.nodes):
            rows = [
                ("id", node.id),
                ("size_bits", str(node.size_bits)),
            ]

            input = [i for i, n in enumerate(self.inputs) if n == node]
            output = [i for i, n in enumerate(self.outputs) if n == node]
            if input:
                rows.append(("input", str(input)))
            if output:
                rows.append(("output", str(output)))

            label = dot_html(dot_table("Node", rows))
            dot.node(f"node-{i}", label, shape="box", color="green")
            for j, input in enumerate(node.inputs):
                head = f"node-{self.nodes.index(input)}"
                tail = f"node-{i}"
                dot.edge(head, tail, headlabel=str(j))

        return dot


@dataclass(frozen=True)
class OperationAllocation:
    id: Optional[str]
    core: Core
    input_memories: Tuple[Memory, ...]
    output_memory: Memory
    time: float
    energy: float


# TODO add set assignment problem rendering to this?
@dataclass(eq=False, frozen=True)
class Problem:
    id: Optional[str]
    hardware: Hardware
    graph: OperationGraph

    possible_allocations: Dict[OperationNode, List[OperationAllocation]]

    # TODO relax this: allow free-to-place nodes, and allow specifying stuff for non-io values
    placement_inputs: Dict[OperationNode, Memory]
    placement_outputs: Dict[OperationNode, Memory]

    def assert_valid(self):
        # TODO assert that all cores are reachable? or at least that there's a single possible solution
        self.hardware.assert_valid()
        self.graph.assert_valid()

        assert set(self.graph.nodes) - set(self.graph.inputs) == self.possible_allocations.keys()
        for node, allocs in self.possible_allocations.items():
            assert len(allocs) > 0

            assert node in self.graph.nodes
            for alloc in allocs:
                assert alloc.core in self.hardware.cores
                assert len(alloc.input_memories) == len(node.inputs)
                for mem in alloc.input_memories:
                    assert mem in alloc.core.connected_memories
                assert alloc.output_memory in alloc.core.connected_memories

        for node, mem in itertools.chain(self.placement_inputs.items(), self.placement_outputs.items()):
            assert node in self.graph.nodes
            assert mem in self.hardware.memories

        assert self.placement_inputs.keys() == set(self.graph.inputs)
        assert self.placement_outputs.keys() == set(self.graph.outputs)

File: rust/src/core/problem.rs
use std::collections::HashSet;

use itertools::Itertools;

use crate::core::wrapper::{Energy, Time, TypedVec};
use crate::define_typed_index;
use crate::util::graphviz::GraphViz;

// problem
#[derive(Debug)]
pub struct Problem {
    pub id: String,
    pub hardware: Hardware,
    pub graph: Graph,

    pub allocations: TypedVec<Allocation, AllocationInfo>,

    pub input_placements: Vec<Memory>,
    pub output_placements: Vec<Memory>,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum CostTarget {
    // Minimize both time and energy, getting the full pareto front
    Full,
    // Primarily minimize time, still optimizing energy as the secondary objective.
    // This will return at most one solution.
    Time,
    // Primarily minimize energy, still optimizing time as the secondary objective.
    // This will return at most one solution.
    Energy,
}

define_typed_index!(Allocation);

// TODO rename this to something better, maybe mapping?
#[derive(Debug)]
pub struct AllocationInfo {
    pub id: String,
    pub group: Group,
    pub node: Node,

    pub input_memories: Vec<Memory>,
    pub output_memory: Memory,

    pub time: Time,
    pub energy: Energy,
}

// graph
#[derive(Debug)]
pub struct Graph {
    pub id: String,
    pub nodes: TypedVec<Node, NodeInfo>,

    pub inputs: Vec<Node>,
    pub outputs: Vec<Node>,
}

define_typed_index!(Node);

#[derive(Debug)]
pub struct NodeInfo {
    pub id: String,
    pub size_bits: u64,
    pub inputs: Vec<Node>,

    // Only start this node after the given set of nodes has been started.
    // This is useful to manually remove symmetries from the optimization problem,
    // making it easier (faster!) to solve.
    pub start_after: Vec<Node>,
}

// hardware
#[derive(Debug)]
pub struct Hardware {
    pub id: String,
    pub memories: TypedVec<Memory, MemoryInfo>,
    pub groups: TypedVec<Group, GroupInfo>,
    pub channels: TypedVec<Channel, ChannelInfo>,
}

// TODO rename group?
define_typed_index!(Group);

define_typed_index!(Channel);

define_typed_index!(Memory);

#[derive(Debug)]
pub struct GroupInfo {
    pub id: String,
}

#[derive(Debug)]
pub struct ChannelInfo {
    pub id: String,
    pub group: Group,
    pub mem_source: Memory,
    pub mem_dest: Memory,
    pub cost: ChannelCost,
}

#[derive(Debug, Copy, Clone, PartialEq)]
pub struct ChannelCost {
    // TODO should we use dedicated allocations per value instead?
    //   this is just halfway limited again
    pub latency: Time,
    pub time_per_bit: Time,
    pub energy_per_bit: Energy,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum ChannelDir {
    Single,
    Both,
}

#[derive(Debug)]
pub struct MemoryInfo {
    pub id: String,
    pub size_bits: Option<u64>,
}

// implementations
impl Problem {
    pub fn assert_valid(&self) {
        self.hardware.assert_valid();
        self.graph.assert_valid();

        for (_, alloc_info) in &self.allocations {
            assert!(self.hardware.groups.has_key(alloc_info.group));
            assert!(self.graph.nodes.has_key(alloc_info.node));
            for &mem in &alloc_info.input_memories {
                assert!(self.hardware.memories.has_key(mem));
            }
            assert!(self.hardware.memories.has_key(alloc_info.output_memory));

            assert_eq!(alloc_info.input_memories.len(), self.graph.nodes[alloc_info.node].inputs.len());
        }

        assert_eq!(self.input_placements.len(), self.graph.inputs.len());
        for &mem in &self.input_placements {
            assert!(self.hardware.memories.has_key(mem));
        }

        assert_eq!(self.output_placements.len(), self.graph.outputs.len());
        for &mem in &self.output_placements {
            assert!(self.hardware.memories.has_key(mem));
        }
    }

    pub fn core_connected_memories(&self) -> Vec<(HashSet<Memory>, HashSet<Memory>)> {
        // TODO this has lost even more meaning
        vec![]
        
        // let mut result = vec![(HashSet::new(), HashSet::new()); self.hardware.cores().len()];
        // 
        // for alloc in &self.allocation_info {
        //     let (inputs, outputs) = &mut result[alloc.core.0];
        //     inputs.extend(alloc.input_memories.iter().copied());
        //     outputs.insert(alloc.output_memory);
        // }
        // 
        // result
    }
}

impl Graph {
    pub fn new(id: impl Into<String>) -> Self {
        Self { id: id.into(), nodes: TypedVec::new(), inputs: vec![], outputs: vec![] }
    }

    pub fn add_node(&mut self, info: NodeInfo) -> Node {
        self.nodes.push(info)
    }

    pub fn create_node(
        &mut self,
        id: impl Into<String>,
        size_bits: u64,
        inputs: Vec<Node>,
        start_after: Vec<Node>,
    ) -> Node {
        let info = NodeInfo {
            id: id.into(),
            size_bits,
            inputs,
            start_after,
        };
        self.add_node(info)
    }

    pub fn add_input(&mut self, node: Node) {
        assert!(self.nodes.has_key(node));
        assert!(self.nodes[node].inputs.is_empty());
        assert!(!self.inputs.contains(&node));
        self.inputs.push(node);
    }

    pub fn add_output(&mut self, node: Node) {
        assert!(self.nodes.has_key(node));
        self.outputs.push(node);
    }

    pub fn to_graphviz(&self) -> GraphViz {
        let mut g = GraphViz::new();

        g.push(format!("label=<<B>{}</B>>", self.id));
        g.push("labelloc=t");

        for (node, node_info) in &self.nodes {
            let mut rows = vec![
                ("debug", format!("{:?}", node)),
                ("id", node_info.id.clone()),
                ("size_bits", node_info.size_bits.to_string()),
            ];

            let input = self.inputs.iter().position(|&x| x == node);
            let output = self.outputs.iter().position(|&x| x == node);
            if let Some(input) = input {
                rows.push(("input", input.to_string()));
            }
            if let Some(output) = output {
                rows.push(("output", output.to_string()));
            }

            let label = GraphViz::table("Node", rows);
            g.push(format!("node_{} [label=<{}>, shape=box, color=green]", node.0, label));

            let tail = format!("node_{}", node.0);

            for (j, input) in node_info.inputs.iter().enumerate() {
                let head = format!("node_{}", input.0);
                g.push(format!("{} -> {} [headlabel={}]", head, tail, j));
            }

            for &after in &node_info.start_after {
                let head = format!("node_{}", after.0);
                g.push(format!("{} -> {} [style=\"dotted\"]", head, tail));
            }
        }

        g
    }

    pub fn assert_valid(&self) {
        // TODO assert that graph is a DAG
        //   even through that should be true by construction, users can still mess around in the public fields

        for (_, info) in &self.nodes {
            for &x in &info.inputs {
                assert!(self.nodes.has_key(x));
            }
            for &x in &info.start_after {
                assert!(self.nodes.has_key(x));
            }
        }

        assert_eq!(self.inputs.iter().unique().count(), self.inputs.len());
        for &x in &self.inputs {
            assert!(self.nodes.has_key(x));
            assert!(self.nodes[x].inputs.is_empty());
        }
        assert_eq!(self.inputs.iter().unique().count(), self.inputs.len());
        for &x in &self.outputs {
            assert!(self.nodes.has_key(x));
        }
    }
}

impl Hardware {
    pub fn new(id: impl Into<String>) -> Self {
        Self { id: id.into(), memories: TypedVec::new(), groups: TypedVec::new(), channels: TypedVec::new() }
    }

    pub fn add_memory(&mut self, info: MemoryInfo) -> Memory {
        self.memories.push(info)
    }

    pub fn create_memory(&mut self, id: impl Into<String>, size_bits: Option<u64>) -> Memory {
        self.add_memory(MemoryInfo { id: id.into(), size_bits })
    }

    pub fn add_group(&mut self, info: GroupInfo) -> Group {
        self.groups.push(info)
    }

    pub fn create_group(&mut self, id: impl Into<String>) -> Group {
        self.add_group(GroupInfo { id: id.into() })
    }
    
    pub fn add_channel(&mut self, info: ChannelInfo) -> Channel {
        let channel = self.channels.push(info);
        self.assert_channel_valid(channel);
        channel
    }

    pub fn create_channel(&mut self, id: impl Into<String>, group: Group, mem_source: Memory, mem_dest: Memory, cost: ChannelCost) {
        let info = ChannelInfo {
            id: id.into(),
            group,
            mem_source,
            mem_dest,
            cost,
        };
        self.add_channel(info);
    }

    pub fn create_channel_bidir(&mut self, id: impl Into<String>, group: Group, mem_left: Memory, mem_right: Memory, cost: ChannelCost) {
        let id = id.into();
        for (dir, mem_source, mem_dest) in [("fwd", mem_left, mem_right), ("bck", mem_right, mem_left)].iter().copied() {
            let info = ChannelInfo {
                id: format!("{}_{}", id, dir),
                group,
                mem_source,
                mem_dest,
                cost,
            };
            self.add_channel(info);
        }
    }

    // TODO render groups as a box around channels + a core if there are allocs in it?
    pub fn to_graphviz(&self, _: Vec<(HashSet<Memory>, HashSet<Memory>)>) -> GraphViz {
        let mut g = GraphViz::new();

        g.push(format!("label=<<B>{}</B>>", self.id));
        g.push("labelloc=t");

        for (mem, mem_info) in &self.memories {
            let size = if let Some(size) = mem_info.size_bits {
                size.to_string()
            } else {
                "inf".to_owned()
            };
            let rows = vec![
                ("debug", format!("{:?}", mem)),
                ("id", mem_info.id.clone()),
                ("size_bits", size),
            ];
            let label = GraphViz::table("Memory", rows);
            g.push(format!("mem_{} [label=<{}>, shape=box, color=blue]", mem.0, label));
        }

        for (channel, channel_info) in &self.channels {
            let rows = vec![
                ("debug", format!("{:?}", channel)),
                ("id", channel_info.id.clone()),
                ("group", self.groups[channel_info.group].id.clone()),
                ("latency", channel_info.cost.latency.0.to_string()),
                ("time_per_bit", channel_info.cost.time_per_bit.0.to_string()),
                ("energy_per_bit", channel_info.cost.energy_per_bit.0.to_string()),
            ];
            let mid = format!("channel_{}", channel.0);

            let label = GraphViz::table("Channel", rows);
            g.push(format!("{} [label=<{}>, shape=box, color=darkorange]", mid, label));

            // TODO fuse pairs of channels that are identical and form a group?
            let head = format!("mem_{}", channel_info.mem_source.0);
            let tail = format!("mem_{}", channel_info.mem_dest.0);

            let dir = "forward";
            g.push(format!("{} -> {} [dir={}, color=darkorange]", head, mid, dir));
            g.push(format!("{} -> {} [dir={}, color=darkorange]", mid, tail, dir));
        }
        
        // TODO render allocs and groups as cores?

        // for  core in self.cores() {
        //     let rows = vec![
        //         ("id", self.core_info[core.0].id.clone()),
        //     ];
        //     let label = GraphViz::table("Core", rows);
        //     g.push(format!("core_{} [label=<{}>, color=green]", core.0, label));
        // 
        //     for (j, mem) in self.memories().enumerate() {
        //         let (inputs, outputs) = &core_connected_memories[core.0];
        // 
        //         let dir = match (inputs.contains(&mem), outputs.contains(&mem)) {
        //             (true, true) => "both",
        //             (true, false) => "input",
        //             (false, true) => "output",
        //             (false, false) => continue,
        //         };
        // 
        //         let head = format!("mem_{}", j);
        //         let tail = format!("core_{}", core.0);
        //         g.push(format!("{} -> {} [dir={}, color=green]", head, tail, dir));
        //     }
        // }

        g
    }

    fn assert_channel_valid(&self, channel: Channel) {
        assert!(self.channels.has_key(channel));
        let channel_info = &self.channels[channel];
        assert!(self.memories.has_key(channel_info.mem_source));
        assert!(self.memories.has_key(channel_info.mem_dest));
    }

    pub fn assert_valid(&self) {
        for channel in self.channels.keys() {
            self.assert_channel_valid(channel);
        }
    }
}

impl ChannelCost {
    pub fn energy_to_transfer(&self, size_bits: u64) -> Energy {
        self.energy_per_bit * size_bits
    }

    pub fn time_to_transfer(&self, size_bits: u64) -> Time {
        self.latency + self.time_per_bit * size_bits
    }
}
